### PR TITLE
Move impala op reg batch 2

### PR DESCRIPTION
**Pull request by matthewmturner**
@datapythonista @jreback 

next batch on the way.  will keep you both posted.

### CODE REVIEW

**Comment by matthewmturner** [comment]
@jreback @datapythonista 

Any idea why CI pyspark test would be failing now? 

i see the `connect` method in `ibis/pyspark/api.py` thats causing the error and `ibis/__init__.py` still doing `import ibis.pyspark.api as pyspark` but we get the following error:

`AttributeError: module 'ibis.pyspark' has no attribute 'connect'`

File: ibis/tests/backends.py [hunk 1]
```diff
@@ -479,7 +479,7 @@ class BigQuery(UnorderedComparator, Backend, RoundAwayFromZero):
 
     @staticmethod
     def connect(data_directory: Path) -> ibis.client.Client:
-        from ibis.bigquery.tests.conftest import connect
+        from ibis.backends.bigquery.tests.conftest import connect
```
Comment (datapythonista) [inline ibis/tests/backends.py:482]: Do you mind changing this in a separate PR? Unfortunately we can't test bigquery in PRs, so I guess this was forgotten when moving the backend. Great to have it fixed, but seems unrelated to this PR.

**Review by datapythonista — COMMENTED** [review]
Changes look good, thanks for working on this.

No idea what can be the problem with pyspark. Can you remove the unrelated change, merge master if this is not updated with it, and then we see if it's still failing? I don't think it should be related, but not the first time we get these weird errors. Hopefully after all this refactoring you're doing we stop having backends break for changes in others.

**Comment by datapythonista** [comment]
> AttributeError: module 'ibis.pyspark' has no attribute 'connect'

There are some weird things happening in the tests. I opened #2478 for something that seems to be the same problem.

File: ibis/tests/backends.py (hunk 1 above)
Comment (matthewmturner) [reply ibis/tests/backends.py:482]: sure - will try again but i had actually only made this update after getting the error (which still hasnt gone away). but yes, i will revert and handle in a separate PR.  would that be its own PR or part of an existing one?

Comment (matthewmturner) [reply ibis/tests/backends.py]: Also, im already up to date with master.

**Comment by matthewmturner** [comment]
> 
> 
> > AttributeError: module 'ibis.pyspark' has no attribute 'connect'
> 
> There are some weird things happening in the tests. I opened #2478 for something that seems to be the same problem.

ok.  i will give that one a shot in the meantime.

File: ibis/tests/backends.py (hunk 1 above)
Comment (datapythonista) [reply ibis/tests/backends.py]: > would that be its own PR or part of an existing one?

I think it's own PR. If you want to check the bigquery build, there may be more errors there, of imports that weren't updated when the backend was moved. I don't think that build was passing before introducing this, so there will be some other unrelated failing tests too.

CC: @tswast

**Comment by datapythonista** [comment]
The spark errors now seem to start to make more sense. I think these operations that you're moving are being imported from the spark backends. Since they can't be found anymore, they'll be generating an `ImportError`. And as you mentioned in a separate comment, these are being suppressed in `ibis/__init__.py`, causing that the attrubite `spark/pyspark` or `ibis` is not set. Which is the error we're getting.

That supression is a really bad practice, since it's not only failing when the third-party dependencies fail, as intended, but also when something is broken. But #2379 will get rid of that, once this refactoring is completed.

**Comment by datapythonista** [comment]
Did you check if we're having some broken imports as I said in my previous comment? I think temporarily removing the suppress should tell you which. I can have a look if you don't have time to check this.

**Comment by matthewmturner** [comment]
yes, sorry i think im still a little confused. We get it here.

```
In [1]: import ibis
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-1-a73c1a9e8fb0> in <module>
----> 1 import ibis

~/development/ibis/ibis/__init__.py in <module>
     61 
     62 # with suppress(ImportError):
---> 63 import ibis.pyspark.api as pyspark  # noqa: F401
     64 
     65 

~/development/ibis/ibis/pyspark/api.py in <module>
----> 1 from ibis.pyspark.client import PySparkClient
      2 from ibis.pyspark.compiler import dialect  # noqa: F401
      3 
      4 
      5 def connect(session):

~/development/ibis/ibis/pyspark/client.py in <module>
----> 1 from pyspark.sql.column import Column
      2 
      3 import ibis.common.exceptions as com
      4 import ibis.expr.types as types
      5 from ibis.backends.spark.client import SparkClient

ModuleNotFoundError: No module named 'pyspark'
```
i made that import local to the class that uses it (`PySparkClient`) - was that the type of change you were expecting?

**Comment by datapythonista** [comment]
> i made that import local to the class that uses it (`PySparkClient`) - was that the type of change you were expecting?

I don't think in this case we should move the import inside the function. First I see that you're not updated with master, the pyspark backend should be inside backends.

Then, the pyspark backend should only be loaded when testing spark. And when testing spark pyspark should be installed. Not sure if this error is local, if it is, can you install the pyspark package, so this doesn't fail, and other problems arise please?

Thanks!

**Comment by matthewmturner** [comment]
I think i found it.  `backends/spark/compiler.py` was using `reduction` from the impala compiler. i updated to from base_sql

**Comment by datapythonista** [comment]
> I think i found it. `backends/spark/compiler.py` was using `reduction` from the impala compiler. i updated to from base_sql

Amazing, looks like there is another error `E   AttributeError: module 'ibis.impala.compiler' has no attribute '_sql_type_names'`, but I think we're close to get a green build.

Thanks for all the work with this, I know it's frustrating having to fix those errors when they are failing silently, but I think we're getting close to get rid of them, once this refactoring is over.

**Comment by matthewmturner** [comment]
yup just saw that and fixed. hopefully that does it.  I made `_sql_type_names` public as well by removing underscore.

**Comment by datapythonista** [comment]
Looks like the CI is still green, the suppress is hiding the error again, you'll have to check was wrong locally.

**Comment by matthewmturner** [comment]
Looks like this is working now. Let me know if any other updates.

**Comment by datapythonista** [comment]
The conda package build error seems to be a genuine error: https://github.com/ibis-project/ibis/pull/2476/checks?check_run_id=1284734683#step:6:1361

Can you have a look please?

**Comment by matthewmturner** [comment]
should be fixed now. isnt it weird that that error was only raised during conda build?  i would have expected SQL tests to fail.

**Comment by matthewmturner** [comment]
i dont see many details on the ibis-project.ibis failed build from azp.  do you have any idea?

**Comment by datapythonista** [comment]
> i dont see many details on the ibis-project.ibis failed build from azp. do you have any idea?

The azure tests are very unreliable, that's why we are moving them to GitHub actions. Your PR is correct, and if you check it, half of the times I rerun the build is an impala job that is failing, and the other half is a windows one. I'll keep restarting until they are green.

**Review by datapythonista — APPROVED** [review]

**Comment by datapythonista** [comment]
Cool, after 4 reruns of azure we've got this green. :)

Thanks for this PR @matthewmturner really nice clean up.

@jreback if you want to have a look...

**Comment by jreback** [comment]
> > i dont see many details on the ibis-project.ibis failed build from azp. do you have any idea?
> 
> 
> 
> The azure tests are very unreliable, that's why we are moving them to GitHub actions. Your PR is correct, and if you check it, half of the times I rerun the build is an impala job that is failing, and the other half is a windows one. I'll keep restarting until they are green.

what exactly is u reliable here?

**Comment by datapythonista** [comment]
> what exactly is unreliable here?

The times I checked, the Windows build was failing because of a database connection error, and the Impala because of conda solver exceeding one hour, the job timeout. We've been having those errors in many PRs, but GitHub actions seems to work well, and I think things will stop failing once we move the builds.

**Comment by jreback** [comment]
> > what exactly is unreliable here?
> 
> 
> 
> The times I checked, the Windows build was failing because of a database connection error, 

sure this is reasonable


> and the Impala because of conda solver exceeding one hour, 

is this not a genuine issue?

**Comment by jreback** [comment]
https://github.com/ibis-project/ibis/runs/1286957259

seems very suspicious - 2 hours to build the recipe?

**Comment by datapythonista** [comment]
> is this not a genuine issue?

There seems to be some randomness in conda, now the build finish in 36 minutes. I'll work in #2480 when I have time, so we can move that build to actions. The problem won't happen anymore in the Impala build, since in actions we only install the required dependencies, not all them. The docs build may be an issue, since it has to install everything so far. But things should improve once the Omnisci backend is out.

In any case, I think we're moving into the right direction to have a CI as simple and reliable as possible, it's just taking a long time because there is a lot of technical debt. And those failures are surely unrelated to this PR, I've seen them in many others.

**Comment by datapythonista** [comment]
> seems very suspicious - 2 hours to build the recipe?

It's the conda solver, it took a long time to do the solving the two times it's doing it. Surely unrelated to this PR, and something that we'll fix by moving omniscidb out. I see other recent PRs also have 90 minutes, or 2 hours. But with the pymapd version change in #2486 it built in 44 minutes, so that seems to be making things better for now.

**Comment by datapythonista** [comment]
I created #2489, we can continue the discussion there. I think this PR can be merged, the problem is not related to the changes here, and the sooner we finish this refactoring, the sooner we'll move omnisci out, and we'll have the problem solved.

**Comment by jreback** [comment]
thanks @matthewmturner really nice.

**Comment by jreback** [comment]
> I created #2489, we can continue the discussion there. I think this PR can be merged, the problem is not related to the changes here, and the sooner we finish this refactoring, the sooner we'll move omnisci out, and we'll have the problem solved.

yep thanks for the review.

**Comment by matthewmturner** [comment]
@datapythonista @jreback 

Trying to figure out which PRs to work on...

Should I prioritize moving the rest of the operation registry for the next ibis release?

**Comment by datapythonista** [comment]
Personally, the things that are top priority to me are:
- Finish the refactoring/decoupling of the backends (mainly moving the operation registry, anything else in other backends that depends on Impala, and finish moving `ibis/sql` to `ibis/backends/sqlalchemy`).
- Move the pending builds to GitHub actions, #2388 and #2389 (both have PRs that don't work yet)

If you don't have a strong preference for anything else, I would pick whatever motivates you the most from those things.

Thanks @matthewmturner

**Comment by matthewmturner** [comment]
Ok sounds good. Will focus on the operation registry for now.

**Comment by datapythonista** [comment]
> Ok sounds good. Will focus on the operation registry for now.

Sounds great. I don't think it should be far from finished, and once it's complete, I think the CI of your PR moving the impala backend should be green. I think we can release a new version at that point. Just couple of file of the `base_sqlalchemy` backend are pending besides that, but that should be very quick to move.
